src/ModelContextProtocol/Client/McpClientExtensions.cs

```diff
@@ -588,6 +582,31 @@ public static Task<ReadResourceResult> ReadResourceAsync(
         return ReadResourceAsync(client, uri.ToString(), cancellationToken);
     }
 
+    /// <summary>
+    /// Reads a resource from the server.
+    /// </summary>
+    /// <param name="client">The client instance used to communicate with the MCP server.</param>
+    /// <param name="uriTemplate">The uri template of the resource.</param>
+    /// <param name="arguments">Arguments to use to format <paramref name="uriTemplate"/>.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="client"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentNullException"><paramref name="uriTemplate"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentException"><paramref name="uriTemplate"/> is empty or composed entirely of whitespace.</exception>
+    public static ValueTask<ReadResourceResult> ReadResourceAsync(
+        this IMcpClient client, string uriTemplate, IReadOnlyDictionary<string, object?> arguments, CancellationToken cancellationToken = default)
+    {
+        Throw.IfNull(client);
+        Throw.IfNullOrWhiteSpace(uriTemplate);
+        Throw.IfNull(arguments);
+
+        return client.SendRequestAsync(
```

Comment:
I'm curious what your bar is for awaiting a Task/ValueTask vs returning it directly. I know it effects exception stack traces, but for small wrappers like this, I personally don't think it matters that much. But then, I saw you added a bunch of async/awaits to some Assert.ThrowsAsync lambdas in our tests, but not here. It couldn't be that it was necessary for the tests to pass, wass it?

Reply:
The async/awaits in the ThrowsAsync lambdas is because they were dealing with ValueTask-returning methods, so they couldn't just be returned.